### ISSUE TITLE
Add easy-to-use integration tests

### ISSUE THREAD

**Issue by fehmer** [paypal/resteasy-spring-boot#46]
It would be nice to give users of this module an easy and nice way to write integration tests for their resteasy jaxrs services.

The tests should 

- not be executed in normal `mvn test` but in `mvn integration-test`
- start the application on a random port
- allow easy definition and validation of requests/responses
- allow to inject mocks in the running application

The documentation should contain all needed information to get started with integration tests to encourage users if this module to write them.

*Edit:* make clear that this issue aims for integration tests for applications which use this module

**Comment by fehmer** [comment]
Hi,

i started working on this issue. I used http://rest-assured.io/ for the definition and validating of the requests. The application is started with SpringApplication.run on a random port and stopped when the test class is finished. The test class is part of the spring java config to allow the definition of mocks.

I added a new module resteasy-spring-boot-test and modified the example-app to use a service for generating the Words. The EchoIT test replaces the word service with a mock and tests the responses.

You can find the code here: https://github.com/fehmer/resteasy-spring-boot/tree/integration-tests. I'll create a PR when i think i am done with my changes.

**Comment by fabiocarvalho777** [comment]
Hello Christian,

Thanks for your contribution, it is definitely valuable.

Just double checking though, I understand this issue would offer to Spring Boot RESTEasy application developers a tool to write integration tests to their applications. But the question is, are you going to develop functional tests for the RESTEasy starter itself too? I am asking because if you are, then that would also tackle issue #13. If that is true, feel free to split your PR in two, so you can work on #13 too. Or you can change the text of this issue send a single PR, and then we close 13 as duplicate.

Thanks!

**Comment by fehmer** [comment]
Hi @fabiocarvalho777,

firstly thanks for releasing this project. I think it is a very good starting point for spring boot and rest easy.

I had a brief look at the open issues to prevent creating a duplicate for this and found #13. I think #13 is meant to introduce functional tests for the starter module itself. My code can be used to write functional tests by creating an application similar to the sample application. This application can make use of as many resteasy features as possible. The integration tests with a running server based on my code will test the functionality of the code from the starter module implicit. Currently I am not planning to tackle #13. I found your project because a friend asks me about spring and resteasy integration. I wrote a similar integration for a client of mine a couple of years ago, but without spring boot. I looked briefly over your implementation and thought that the integration test part is missing from the documentation.

Long story short, I just want to provide the testing module and some documentation/examples to encourage the users of your module to write integration tests. If I use your module myself in the near future i might write some tests for your module using an example application as described above.

I am preparing the PR and looking forward your acceptance or a further discussion.

**Comment by fehmer** [comment]
PR: #47

**Comment by fabiocarvalho777** [comment]
Ok, after taking a better look at this suggestion and its PR, I would like to discuss it more.

I have noticed that there is nothing on its concept and implementation that is really RESTEasy specific. It could be used pretty much against any REST application developed using Spring Boot, even if it uses Jersey instead of RESTEasy, or even a non JAX-RS solution, like Spring Boot REST apps using Spring MVC.

Because of that, although valuable, it might not be really part of the scope of this starter. Maybe a better place for it would be under REST-assured, or Spring Boot. Also, it is important to me to keep the scope of this starter focused, that will help to maintain it.

It would be nice too if it supports more than one single unit test framework (TestNG for example).

Regardless of that, I can totally see the code in your PR being adapted to address #13.

What are your thoughts about all of it?

Thanks.

**Comment by fehmer** [comment]
Hi @fabiocarvalho777,

you are right. The implementation is not RESTEasy specific at all. The concept is to test the rest service as a client and the client does not know that the server is implemented using RESTEasy.

The new module has no dependency to the resteasy-spring-boot modules at all, so it might be the wrong place. On the other hand it would be nice to provide the users of your starter a way to write integration tests. Should we provide the code and the maven pom changes as documentation only and not release a new maven module? I am not aware if somewhere a similar module is available. If so,  including it in your documentation and the sample-app would solve the issue/improvement. An other way would be to move the RestEasyApplicationTest to the sample-app. 

About the test frameworks. RestEasyApplicationTest just ties together SocketUtils from Spring, a Spring Context and configs RestAssured. The non-junit related code is like 3 lines and can surely be adapted to any test framework you like. The sample app uses junit, spring-boot-starter-test uses junit and I personally use junit too. So i used junit here. The intention of starters is IMO to provide one, opinionated way to do something. If someone want to use testNG they have to get rid of the resteasy-spring-boot-test and spring-boot-starter-test.

Feel free to use the code/concept for #13.

I understand your opinion that the starter isn't the 100% right place for that. I don't want to discuss the scope of your project or your opinion, its your choice. I think providing the code to the users of the starter and encourage them to write integration tests would be nice. They probably won't mind that org.springframework:rest-boot-test-starter would be a better dependency. 

What do you think, should we reject the PR completely. Should we move the base class to the sample-app and only write some documentation about this?

I will stop working on the #47 for now until we know where to put it. 

Just two things about your comments in the code:

- In the `develop` branch the echo method is located at `/sample-app/echo/echo`. So the basepath of the service tested by the EchoIT is `/sample-app/echo`. This matches with the `@Path` on the Echo class. 

- The is a problem with the spring context in the code i will have to fix. Parallel executions on class level would be possible with some changes. Parallel executions on method level would need major changes in the way the spring context is created and cause the application to get started on each test method. What do you think, should i restrict parallel executions to class level?

Thank you for your feedback.

Christian

**Comment by fabiocarvalho777** [comment]
Hi Christian,

I prefer to keep the scope of this project strict, that is important to make it maintainable. So I prefer to not add the test framework, but thanks for your contribution.

Like I said earlier, I still think most of that code can be used as solution for #13 though. So, if you want, you can convert the code in PR 47 accordingly and use it for #13. Or we can just close it and create a new one specifically for #13.

Thanks.

**Comment by fehmer** [comment]
okay, close the PR please.
